### ISSUE TITLE
useFonts hook randomly failing to load fonts

### ISSUE THREAD

**Issue by crschmidt2** [expo/expo#21885]
### Summary

When I use the {useFonts, Abel_400Regular} hook from "@expo-google-fonts/abel" and {HankenGrotesk_300} from "@expo-google-fonts/hanken-grotesk" the Abe font will fail to load on my iOS Expo Go app and my Android studio emulator. However, when I add the function 
```
const onLayoutRootView = useCallback(async () => {
		if (fontsLoaded) {
		  await SplashScreen.hideAsync();
		}
	  }, [fontsLoaded]);
```

and save my editor the font loads fine. But when I reload each app from Expo Cli the font refuses to load again unless I remove the function, save in my editor, add the function, and save again. I had this issue earlier with locally saved .otf font files, so I believe the issue is with the { useFonts } hook. I've tried all fixes I can find online and nothing's helped.

EDIT: I found out it's loading the fonts whenever I hot-reload after changing any code in the render return statement for a function component. I just added an if (fontsLoaded) {} and saved and it loaded the fonts. I'm guessing it's throwing the Font not loaded error slightly before the font gets a chance to load, but when I use a splashscreen to wait for it to load it gets stuck on the splashscreen forever.

### What platform(s) does this occur on?

Android, iOS

### SDK Version

~48.0.9

### Environment

```
expo-env-info 1.0.5 environment info:
    System:
      OS: Windows 10 10.0.19044
    Binaries:
      Node: 18.13.0 - C:\Program Files\nodejs\node.EXE
      npm: 9.6.0 - C:\Program Files\nodejs\npm.CMD
    IDEs:
      Android Studio: AI-221.6008.13.2211.9619390
    npmPackages:
      @expo/webpack-config: ^18.0.1 => 18.0.2
      expo: ~48.0.9 => 48.0.9
      react: 18.2.0 => 18.2.0
      react-dom: 18.2.0 => 18.2.0
      react-native: 0.71.4 => 0.71.4
      react-native-web: ~0.18.10 => 0.18.12
    Expo Workflow: managed
```

### Minimal reproducible example

Here is the minimum code I was able to reproduce the error with. When I change the return statement and hot-save the project, the render update fixes the font. 
```
import React, { useCallback } from "react";
import {StyleSheet, Text} from 'react-native'
import { NavigationContainer } from "@react-navigation/native";
import { createNativeStackNavigator } from "@react-navigation/native-stack";
import { useFonts, Abel_400Regular } from "@expo-google-fonts/abel";
import { HankenGrotesk_300Light } from "@expo-google-fonts/hanken-grotesk";
import * as SplashScreen from "expo-splash-screen";

import HomeScreen from "./screens/HomeScreen";
import ActiveSubscriptionsScreen from "./screens/ActiveSubscriptionsScreen";
import CustomNavigationBar from "./components/CustomNavigationBar";
import LogInScreen from "./screens/LoginScreen";
import ViewOrdersScreen from "./screens/ViewOrdersScreen";
import CartScreen from "./screens/CartScreen";
import AccountScreen from "./screens/AccountScreen";
import SignUpScreen from "./screens/SignUpScreen";



export default function App() {
	const [fontsLoaded] = useFonts({
		Abel_400Regular,
		HankenGrotesk_300Light
	})

	const onLayoutRootView = useCallback(async () => {
		if (fontsLoaded) {
		  await SplashScreen.hideAsync();
		}
	  }, [fontsLoaded]);

	return (
		<Text style={{fontFamily: "Abel_400Regular", alignSelf: "center", marginTop: 500}}>Styling! :D Hooray!</Text>
	);
}

```
Here is one of the screens the font is running on
```

**Comment by robert-nash** [comment]
I am also facing a similar issue

**Comment by crschmidt2** [comment]
Robert: I actually figured out a fix that worked for me. For some reason the splashScreen would never register the value of fontsLoaded changing, but I was able to fix the issue by using the deprecated element called AppLoading from expo-app-loading. All I did was do 
`
if (!fontsLoaded) {
    return <AppLoading/>
}
`
This correctly kept the screen loading until the fonts were loaded. After they were loaded it stopped showing the loading screen. Hope this helps you out, and if anybody figures out why it's not working with SplashScreen please let me know

**Comment by douglowder** [comment]
I took a look at this and I can reproduce this reliably on Android when downloading an EAS update and reloading the app with the update. Investigating.

**Comment by douglowder** [comment]
ENG-9592

**Comment by douglowder** [comment]
@crschmidt2 you are correct that you should show a loading component and not null while fonts are not loaded. Also, we recommend that you handle errors from the `useFonts()` API to prevent the splash screen from showing forever:

```jsx
export default function App() {
	const [fontsLoaded, fontError] = useFonts({
		Abel_400Regular,
		HankenGrotesk_300Light
	})

	const onLayoutRootView = useCallback(async () => {
		if (fontsLoaded || fontError) {
		  await SplashScreen.hideAsync();
		}
	  }, [fontsLoaded, fontError]);

        if (!fontsLoaded && !fontError) { return <AppLoading/> }

	return (
		<Text style={{fontFamily: "Abel_400Regular", alignSelf: "center", marginTop: 500}}>Styling! :D Hooray!</Text>
	);
}
```

**Comment by mphill** [comment]
`useFont` will randomly return false for loaded.  It appears to be completely random.  There must be some loading timing or thread issue

Seeing this in Expo 49 currently, if you implemented `useFont` per the documentation:

```ts
  const onLayoutRootView = useCallback(async () => {
    if (fontsLoaded) {
      await SplashScreen.hideAsync();
    }
  }, [fontsLoaded]);

  if (!fontsLoaded) {
    return null;
  } 
```

Your app will randomly hang on the splash screen on iOS.  I haven't confirmed on Android.  It's easy to reproduce this. Force quit 10-15 times and you will get the app to hang on the splashscreen.

I also added a `setTimeout` to force the splash to hide because this is a pretty bad issue to have out in the wild.  It possible a fresh install will never get into the app, I've seen this a few times now.

**Comment by carbonatedcoder** [comment]
@douglowder, in your example, you recommend proceeding into the app even if there is a `fontError`. If the font truly did not load and there was a `fontError`, is entering the app truly the recommended course of action? Won't that cause a flood of errors that look like `fontFamily "[font_name]" is not a system font and has not been loaded through Font.loadAsync.` in any component that uses the custom font (which are presumably many)?

We are also seeing the same behavior described by @mphill  in Expo SDK 49. We are using the `useFonts` hook from within `@expo-google-fonts/montserrat` version `0.2.3`. It seems to be completely random, but is reproducible if you try to force close and restart the app enough times. Since `fontsLoaded` is returning false, the app component never mounts and its `onLayout` callback never called, which is what should be hiding the splash screen.

**Comment by douglowder** [comment]
You make a good point -- depending on the app, it may be desirable to show an error screen or have code that handles falling back to system fonts.

My main concern was that the splash screen be hidden even if errors occur, otherwise font problems will be very difficult to debug.

**Comment by mphill** [comment]
For what it’s worth I don’t think this is the fonts hook that is failing, it's a assets loading issue.  When this issue shows up, the in-app images don't load either.

**Comment by carbonatedcoder** [comment]
> For what it’s worth I don’t think this is the fonts hook that is failing, it's a assets loading issue. When this issue shows up, the in-app images don't load either.

This does seem to be correct or at least related. In the errors we are seeing reported, it is complaining that the font asset file doesn't even exist. 

```
File 'file:///var/mobile/Containers/Data/Application/[....]/Library/Application%20Support/.expo-internal/assets/node_modules/@expo-google-fonts/montserrat/Montserrat_400Regular.ttf' for font 'Montserrat_400Regular' doesn...
```

Given the version that is being logged, it looks like the issue is also causing the app to fallback to the bundle that was shipped with the binary build versus the latest OTA. This seems to be a pretty prevalent issue for us, so any additional findings would be welcome.

**Comment by carbonatedcoder** [comment]
@mphill have you found any usable workaround in the meantime? We are seeing the issue roughly every 10-15 app starts, which is pretty rough. Given that the underlaying issue appears to be something with the loading of assets (i.e. the font file doesn't even appear available to the app), we can't necessarily hide the splash screen (i.e. `setTimeout` method) and allow the user to proceed into the app if the font isn't there.

The issue appears to be temporarily fixed by a force quit + reopen of the app. But, it can return again randomly at any point.

Given that, we were going to programmatically restart the app with `Updates.reloadAsync` for the user upon failure, but that doesn't even seem to work either as it raises a separate exception, despite being successfully used elsewhere in the project and having updates enabled: `Calling the 'reload' function has failed Caused by: Cannot call module method when expo-updates is disabled`

**Comment by douglowder** [comment]
@carbonatedcoder if you are seeing the message above, it means that you are not actually using `expo-updates`, and should remove it from your app.

**Comment by carbonatedcoder** [comment]
> if you are seeing the message above, it means that you are not actually using `expo-updates`, and should remove it from your app.

@douglowder, thanks for the followup, but we do actually make use of `expo-updates` in the app and have a custom handler for OTA update checking/processing (`checkForUpdateAsync`, `fetchUpdateAsync`, and `reloadAsync`). 

Additionally, we use `reloadAsync` in a few places throughout the app successfully to restart the app for different purposes. The only time it fails with this error is while being called during/after this assets issue is taking place. So, I'm not sure if it's something to do with the app falling back to the built in bundle or what.

**Comment by karamjb** [comment]
Hello,
I am using expo update 
and I got this error message

```File 'file:///var/mobile/Containers/Data/Application/*****uuid/Library/Application%20Support/.expo-internal/assets/assets/fonts/Somar-Bold.otf' for font 'Somar-Bold' doesn't exist ```

**Comment by douglowder** [comment]
@carbonatedcoder if you are going to use expo-updates, it needs to be configured to use EAS update server, or your own server that follows our updates protocol.

Could you share the contents of `ios/<appname>/Supporting/Expo.plist`?

**Comment by carbonatedcoder** [comment]
@douglowder, this is a managed workflow project so it's all default configuration when it comes to EAS. No custom servers are specified and we've used this for months without issue after switching from classic updates to EAS. We've never had issues with it.

The OTA updates are checked and download just fine, so the issue is not really with that process. The only reason I called out the updates exception is because our idea was to reload the app for the user but this exception is *only* seen when trying to call `reloadAsync` while this issue with the assets (font issue) is occurring. That error is not thrown anywhere else in our updates flow, which makes use of `reloadAsync`, elsewhere it just works.

The custom Updates strategy we use is similar to the one described [here](https://docs.expo.dev/versions/latest/sdk/updates/), with the addition of redirecting the user to a different screen while the update is happening. The `updates.checkAutomatically` key is set to `ON_ERROR_RECOVERY` so that the custom strategy is used.

**Comment by douglowder** [comment]
Ok so this

> `Calling the 'reload' function has failed Caused by: Cannot call module method when expo-updates is disabled`

is only happening at startup, when you detect the font loading issue?

**Comment by carbonatedcoder** [comment]
> Ok so this
> 
> > `Calling the 'reload' function has failed Caused by: Cannot call module method when expo-updates is disabled`
> 
> is only happening at startup, when you detect the font loading issue?

That's correct. We can (and do) call `reloadAsync` in multiple other places in the app just fine. It's only when this assets/font issue occurs at startup does a then manual call to `reloadAsync` throw the updates related error that is not seen elsewhere.

Until the root cause is fixed, we thought we could mask the issue (to the user) by using `Updates.reloadAsync` when a `fontError` (from useFonts) occurred, but quickly found it wouldn't work due to the mentioned issue.

All of these somehow stem from the assets/font issue where, completely randomly, the app will start and be unable to locate the asset/font file. So whatever is happening there is likely related to this other behavior. We see it maybe once per 10-15 app starts.

**Comment by douglowder** [comment]
Thanks for this information. Now I suspect that you are calling `reloadAsync` while the updates controller is still starting up. To test this, change your code to issue the `reloadAsync` with a delay of a second or two.

I'm attempting to set up a repro for your issue on iOS with SDK 49.

**Comment by carbonatedcoder** [comment]
> Thanks for this information. Now I suspect that you are calling `reloadAsync` while the updates controller is still starting up. To test this, change your code to issue the `reloadAsync` with a delay of a second or two.

No problem! Oddly enough, we were actually already delaying the `reloadAsync` call with a delay of ~5 seconds just in case, so I would presume that should be enough time for things to settle out.

Strangely enough, when the issue is encountered and the errors get logged, the version reported in the Sentry event is the one from binary build time, despite hitting code that was only added via a subsequent OTA in a later update. So, given that, my one remaining theory was that maybe Expo is detecting the error on/during startup and is trying to fallback to an embedded version or something. But, even that version should have the assets.

**Comment by douglowder** [comment]
@carbonatedcoder does the issue ever happen on a device that has only the embedded assets (has never downloaded and launched an OTA update)?

**Comment by carbonatedcoder** [comment]
> @carbonatedcoder does the issue ever happen on a device that has only the embedded assets (has never downloaded and launched an OTA update)?

This issue only started happening for us when we upgraded from SDK 48 -> 49. We started seeing reports of this issue fairly soon after the new build was published, even while it was still only using the embedded bundle. In other words, the issue was occurring before we even published an OTA under 49. It has continued even after OTAs have been pushed out.

**Comment by douglowder** [comment]
Thanks much! It would be useful to know the versions you are using of the following packages:

- expo-updates
- expo-asset
- expo-font
- expo-manifest

**Comment by carbonatedcoder** [comment]
> Thanks much! It would be useful to know the versions you are using of the following packages:

- `expo-updates` 0.18.13 
- `expo-asset` 8.10.1 (dependency via `expo`)
- `expo-font` 11.4.0 (dependency via `expo`)
- `expo-manifests` 0.7.2 (dependency via `expo-updates`)

**Comment by douglowder** [comment]
Ok @carbonatedcoder now testing release builds of this app:

```jsx
import React, { useEffect } from 'react';
import { Text, SafeAreaView } from 'react-native';
import { useFonts, Abel_400Regular } from '@expo-google-fonts/abel';
import { HankenGrotesk_300Light } from '@expo-google-fonts/hanken-grotesk';
import * as SplashScreen from 'expo-splash-screen';

SplashScreen.preventAutoHideAsync();

export default function App() {
  const [fontsLoaded, error] = useFonts({
    Abel_400Regular,
    HankenGrotesk_300Light,
  });

  useEffect(() => {
    if (fontsLoaded || error) {
      SplashScreen.hideAsync();
    }
  }, [fontsLoaded, error]);

  return (
    <SafeAreaView style={{ flex: 1, justifyContent: 'center', alignItems: 'center' }}>
      <Text style={{ fontFamily: 'Abel_400Regular', alignSelf: 'center', marginTop: 500 }}>
        Styling! :D Hooray!
      </Text>
      <Text>{`Error: ${error}`}</Text>
    </SafeAreaView>
  );
}
```

**Comment by douglowder** [comment]
Ah the weird behavior is a rendering issue, due to the fact that the style object could get used before or after `fontsLoaded` is set. If I write the component as shown below, then I always see correctly styled text. I do not reproduce any font loading error.

```tsx
const textStyle = fontsLoaded ? [styles.text, { fontFamily: 'Abel_400Regular' }] : styles.text;

  return (
    <SafeAreaView style={styles.container}>
      <View style={{ flex: 1 }} />
      {fontsLoaded ? (
        <Text style={textStyle}>Styling! :D Hooray!</Text>
      ) : (
        <Text style={textStyle}>{`Error: ${error}`}</Text>
      )}
      <Pressable onPress={handleReload} style={styles.button}>
        <Text style={styles.buttonText}>Reload</Text>
      </Pressable>
      <Text style={styles.text}>{`Reload error: ${reloadError ?? ''}`}</Text>
      <View style={{ flex: 1 }} />
    </SafeAreaView>
  );
```

**Comment by karamjb** [comment]
Did you try rapid force close and open? its not reproducible in ExpoGo app

**Comment by carbonatedcoder** [comment]
@douglowder, thank you for attempting to repro this issue.

Just a couple additional points:
- We only notice this issue in production builds. It has not be reproduced in Expo Go.
- We haven't tried to actually restart the app using `reloadAsync` manually from a button like that to trigger the behavior. It's most notably seen by opening the app, force quitting it from the iOS app switcher, then reopening the app again after a few seconds. Within ~10 times, we are able to repro it.

**Comment by douglowder** [comment]
@karamjb @carbonatedcoder yes I rapidly force closed and opened the app -- no font errors were seen. This is a release standalone build, not Expo Go.

**Comment by douglowder** [comment]
I've created a repo of my test app for this at https://github.com/douglowder/UpdatesFonts .

**Comment by Ayman-Yahia** [comment]
Hi 
i am facing the same issue here as if there no assets file to import from.
Is there any workaround to fix it?

**Comment by sevketaydogdu** [comment]
I am also encountering the same problem. 
I set the font loading state to true with settimeout and the application opens like that.

**Comment by coffenbacher** [comment]
I am also running into this, I stripped my App.tsx down to practically nothing to see if something being executed in parallel with the fonts was somehow triggering this and I still hit this. It feels like a race condition or something like that in a core boot-up process.

**Comment by douglowder** [comment]
> I've created a repo of my test app for this at https://github.com/douglowder/UpdatesFonts .

@coffenbacher @carbonatedcoder I was not able to reproduce this error with the above test app. It seems like @coffenbacher might have a stripped down `App.tsx` that does reproduce the problem. Could you share a repo with the project that shows the problem when the app is force quit and restarted?

**Comment by carbonatedcoder** [comment]
After days of debugging and countless builds, we found the problem, at least in our case. The culprit, out of all things, was `expo-insights`. The app was stripped bare to a single, basic component and when that still didn't work, we started removing non-essential packages until we found it.

Something within `expo-updates` is breaking the app startup process. In our case, when the behavior was encountered on cold app starts, it would prevent any assets (fonts, images) from being loaded *and* tell us that `expo-updates` was disabled, when it was not. After removing insights and rebuilding, we are no longer seeing the issue loading assets. This was incredibly hard to debug because there were never any errors raised even hinting that it could be related to `expo-insights`.

The version of `expo-insights` installed was `0.2.0`. For what it's worth, this was likely an iOS only issue. @douglowder, you may want to relay these findings to the `expo-insights` team, if you are able to.

⚠️  **For anyone experiencing this issue, check if you have `expo-insights` installed and, if so, remove it and rebuild your app.** ⚠️

**Comment by sevketaydogdu** [comment]
> After days of debugging and countless builds, we found the problem, at least in our case. The culprit, out of all things, was `expo-insights`. The app was stripped bare to a single, basic component and when that still didn't work, we started removing non-essential packages until we found it.
> 
> 
> 
> Something within `expo-updates` is breaking the app startup process. In our case, when the behavior was encountered on cold app starts, it would prevent any assets (fonts, images) from being loaded *and* tell us that `expo-updates` was disabled, when it was not. After removing insights and rebuilding, we are no longer seeing the issue loading assets. This was incredibly hard to debug because there were never any errors raised even hinting that it could be related to `expo-insights`.
> 
> 
> 
> The version of `expo-insights` installed was `0.2.0`. For what it's worth, this was likely an iOS only issue. @douglowder, you may want to relay these findings to the `expo-insights` team, if you are able to.
> 
> 
> 
> ⚠️  **For anyone experiencing this issue, check if you have `expo-insights` installed and, if so, remove it and rebuild your app.** ⚠️  
> 
> 
> 
> 

Thank you so much i will try to remove this package

**Comment by douglowder** [comment]
Thank you thank you @carbonatedcoder ! We will investigate and determine why `expo-insights` is causing this issue.

**Comment by douglowder** [comment]
@carbonatedcoder FYI we have implemented a change that we believe will fix the issue, in #24675 . The test app I am using no longer shows the error after this change. If you wish, you can add `expo-insights` back to your app, and use `patch-package` to make that change in the module, and then build as usual with the managed workflow, and see if the error still occurs in rare cases. Thanks again for your help in finding this!

**Comment by carbonatedcoder** [comment]
@douglowder, great, thanks! Glad that seemed to be a relatively quick fix for `expo-insights`. Happy to have helped narrow that one down! It was truly stumping us for quite awhile.

**Comment by mphill** [comment]
@carbonatedcoder great work.

**Comment by RyanTG** [comment]
Oh my goodness, we had been tearing our hairs out for weeks trying to resolve this issue. Very hard to diagnose. Can't believe it was Expo-Insights this whole time. Sorry, ol' package, we cannot risk something like this again.

Thank you so much for figuring out this issue.

**Comment by nathantew14** [comment]
I seem to be getting this on android, and only in production builds on physical devices. Sometimes my app will hang with a blank screen, after the splashscreen is hidden:

```
const onLayoutRootView = useCallback(async () => {
    if (fontsLoaded) {
        await SplashScreen.hideAsync();
    }
}, [fontsLoaded]);

Other times it hangs on the splashscreen itself. Nothing seems to be reproducible reliably.

**Comment by nathantew14** [comment]
@douglowder I actually tried to capture the error with Sentry and automatically restart the app on `fontError`, but for some reason the error is never captured by sentry and restarting doesn’t resolve the issue

in your suggestion you just hid the splashscreen upon error (which is essentially ignoring it, right?), but from the comments above it seems like a failure to load the fonts would indicate a failure to load all other assets?

**Comment by douglowder** [comment]
@nathantew14 I am not suggesting that the error be ignored -- it does look like a failure to load all assets. I was just saying that it will help with debugging if you hide the splash screen when the error happens. In cases like this, I display the error in the UI as well.

If you hide the splashscreen and then have a useEffect that throws when the error happens, I think Sentry should hopefully be able to capture that.

**Comment by brunops** [comment]
Hi, is there a solution for this? Expo app works fine with in Expo Go, but as soon as I make an ios build, the app in TestFlight freezes in the Splash Screen, not sure how debug this.

thanks

**Comment by cs-manughian** [comment]
I'm not using expo-insights and I'm still seeing this issue - any other solutions out there?

**Comment by tkoehlerlg** [comment]
@douglowder

**Comment by itsezc** [comment]
Tried all solutions suggested here, including loading from `@expo-google-fonts` - no luck on Expo 49. My CLI is throwing errors about f This issue should remain open @douglowder.

**Comment by AlbertoAquinoDev** [comment]
Happens the same to me right now...

**Comment by aindong** [comment]
This issue still happens on sdk 50

**Comment by douglowder** [comment]
@itsezc @AlbertoAquinoDev @aindong do you have an Expo SDK 50 project that shows this issue? If so, please provide that and the steps you follow when observing the font loading bug (including whether the bug happens on iOS, Android, or both). If you can provide that, I'll create a new GitHub issue (as it is probably not the same as the problem originally happening in this issue).

**Comment by DutchPrince** [comment]
Same problem here using Expo SDK 50, did the default expo install with @expo-google-fonts/archivo and i also get the warning "fontFamily "Archivo500Medium" is not a system font and has not been loaded through expo-font." inside my pages that use the below root layout. Also the same error when loading the fonts from assets. Feel like i tried everything, resetting my project, clearing ./node_modules/expo/node_modules/expo-font/, rebuilding pods etc.

```
import * as SplashScreen from 'expo-splash-screen';
import {
    useFonts,
    Archivo_100Thin,
    Archivo_200ExtraLight,
    Archivo_300Light,
    Archivo_400Regular,
    Archivo_500Medium,
    Archivo_600SemiBold,
    Archivo_700Bold,
    Archivo_800ExtraBold,
    Archivo_900Black,
    Archivo_100Thin_Italic,
    Archivo_200ExtraLight_Italic,
    Archivo_300Light_Italic,
    Archivo_400Regular_Italic,
    Archivo_500Medium_Italic,
    Archivo_600SemiBold_Italic,
    Archivo_700Bold_Italic,
    Archivo_800ExtraBold_Italic,
    Archivo_900Black_Italic,
} from '@expo-google-fonts/archivo';


function RootLayout() {
    let [fontsLoaded, error] = useFonts({
        'Archivo100Thin': Archivo_100Thin,
        'Archivo200ExtraLight': Archivo_200ExtraLight,
        'Archivo300Light': Archivo_300Light,
        'Archivo400Regular': Archivo_400Regular,
        'Archivo500Medium': Archivo_500Medium,
        'Archivo600SemiBold': Archivo_600SemiBold,
        'Archivo700Bold': Archivo_700Bold,
        'Archivo800ExtraBold': Archivo_800ExtraBold,
        'Archivo900Black': Archivo_900Black,
        'Archivo100ThinItalic': Archivo_100Thin_Italic,
        'Archivo200ExtraLightItalic': Archivo_200ExtraLight_Italic,
        'Archivo300LightItalic': Archivo_300Light_Italic,
        'Archivo400RegularItalic': Archivo_400Regular_Italic,
        'Archivo500MediumItalic': Archivo_500Medium_Italic,
        'Archivo600SemiBoldItalic': Archivo_600SemiBold_Italic,
        'Archivo700BoldItalic': Archivo_700Bold_Italic,
        'Archivo800ExtraBoldItalic': Archivo_800ExtraBold_Italic,
        'Archivo900BlackItalic': Archivo_900Black_Italic,
    });

    const onLayoutRootView = useCallback(async () => {
        if (fontsLoaded && !error) {
            await SplashScreen.hideAsync();
        }
    }, [fontsLoaded, error]);

    if (!fontsLoaded && !error) {
        return null;
    }
    
    
    return (
                <View onLayout={onLayoutRootView} style={{
                    flex: 1,
                }}>
                   //
                </View>
    )
}

export default RootLayout;
```

**Comment by DutchPrince** [comment]
For me it works inside the RootLayout but not inside the Stack component thats loaded within 

`
import * as SplashScreen from 'expo-splash-screen';

import {
    useFonts,
    Archivo_100Thin,
    Archivo_200ExtraLight,
    Archivo_300Light,
    Archivo_400Regular,
    Archivo_500Medium,
    Archivo_600SemiBold,
    Archivo_700Bold,
    Archivo_800ExtraBold,
    Archivo_900Black,
    Archivo_100Thin_Italic,
    Archivo_200ExtraLight_Italic,
    Archivo_300Light_Italic,
    Archivo_400Regular_Italic,
    Archivo_500Medium_Italic,
    Archivo_600SemiBold_Italic,
    Archivo_700Bold_Italic,
    Archivo_800ExtraBold_Italic,
    Archivo_900Black_Italic,
} from '@expo-google-fonts/archivo';


function RootLayout() {
    let [fontsLoaded, error] = useFonts({
        'Archivo100Thin': Archivo_100Thin,
        'Archivo200ExtraLight': Archivo_200ExtraLight,
        'Archivo300Light': Archivo_300Light,
        'Archivo400Regular': Archivo_400Regular,
        'Archivo500Medium': Archivo_500Medium,
        'Archivo600SemiBold': Archivo_600SemiBold,
        'Archivo700Bold': Archivo_700Bold,
        'Archivo800ExtraBold': Archivo_800ExtraBold,
        'Archivo900Black': Archivo_900Black,
        'Archivo100ThinItalic': Archivo_100Thin_Italic,
        'Archivo200ExtraLightItalic': Archivo_200ExtraLight_Italic,
        'Archivo300LightItalic': Archivo_300Light_Italic,
        'Archivo400RegularItalic': Archivo_400Regular_Italic,
        'Archivo500MediumItalic': Archivo_500Medium_Italic,
        'Archivo600SemiBoldItalic': Archivo_600SemiBold_Italic,
        'Archivo700BoldItalic': Archivo_700Bold_Italic,
        'Archivo800ExtraBoldItalic': Archivo_800ExtraBold_Italic,
        'Archivo900BlackItalic': Archivo_900Black_Italic,
    });

    const onLayoutRootView = useCallback(async () => {
        if (fontsLoaded && !error) {
            await SplashScreen.hideAsync();
        }
    }, [fontsLoaded, error]);

    if (!fontsLoaded && !error) {
        return null;
    }

    return (
                <View onLayout={onLayoutRootView} style={{
                    flex: 1,
                }}>
                                <Stack screenOptions={{
                                    headerShown: false,
                                }} />
                     
                </View>
    )
}

export default RootLayout;`

**Comment by parkwherever** [comment]
@douglowder I got this issue straight off the bat with `npx create-expo-app@latest -t tabs@50`

**Comment by parkwherever** [comment]
It's from the root layout and I am seeing it only with the local font 
```
  const [loaded, error] = useFonts({
    SpaceMono: require("../assets/fonts/SpaceMono-Regular.ttf"),
    ...FontAwesome.font,
  });
```
though the icons also do not load (albeit not on the load path)

Commenting out SpaceMono enables the load of the FontAwesome font

**Comment by douglowder** [comment]
@parkwherever thanks for the report. This seems like a different issue than the one originally posted here... is the font error happening every time? Please provide a repo with the code and assets to repro the issue.

**Comment by ganeshkumarburra** [comment]
Hi @douglowder 
Any input on above issue, fonts are loading outside of the Navigation Container/Stack Screens.
But when we are using the fonts in the Screen components fonts are not working as expected.

**App.js File**
```js
import React, { useCallback, useEffect, useState } from "react";
import { NavigationContainer } from "@react-navigation/native";
import { createNativeStackNavigator } from "@react-navigation/native-stack";
import { View } from "react-native";
import { ThemeProvider } from "@shopify/restyle";
import Login from "./src/components/Authentication/Login";
import theme from "./src/theme";
import Signup from "./src/components/Authentication/Signup";
import { useFonts } from "expo-font";
import * as SplashScreen from 'expo-splash-screen';

const Stack = createNativeStackNavigator();

export default function App() {
  const [fontsLoaded, fontError] = useFonts({
    'LemonLove': require('./assets/fonts/LemonLove.ttf')
  });

  const onLayoutRootView = useCallback(async () => {
    if (fontsLoaded || fontError) {
      await SplashScreen.hideAsync();
    }
  }, [fontsLoaded, fontError]);

  if (!fontsLoaded && !fontError) {
    return <View>
      <Text>Fonts not loaded</Text>
    </View>
  }

  return (
    <ThemeProvider theme={theme}>
      <Text style={{  fontSize: 30, fontFamily: 'LemonLove' }}>Inter Black 3</Text>
      <NavigationContainer>
        <Stack.Navigator>
          <Stack.Screen
            name="login"
            component={Login}
            options={{ headerShown: false }}
          />
          <Stack.Screen
            name="register"
            component={Signup}
            options={{ headerShown: false }}
          />
        </Stack.Navigator>
      </NavigationContainer>
    </ThemeProvider>
  );
}

```
**Signup Component:** 
```js
import React from "react";
import { Box, SafeTopSpace, ScrollView, Text } from "../Common";
import { StyleSheet } from "react-native";

const Signup = () => {
  const styles = StyleSheet.create({
    heading: {
      fontFamily: "LemonLove",
      fontSize: 28,
      fontWeight: "700",
      lineHeight: 30,
      textAlign: "left",
    },
  });
  return (
    <Box
      flex={1}
      backgroundColor={"grayB"}
      height={500}
      paddingVertical={"m"}
      paddingHorizontal={"s"}
    >
      <SafeTopSpace />
      <Text color={"grayA"} style={styles.heading}>
        Lets Create a new Account
      </Text>
      <Text color={"grayC"}>Please fill in the data below</Text>
      <Box
        padding={"s"}
        marginVertical={"s"}
        borderRadius={"medium"}
        backgroundColor={"white"}
      ></Box>
    </Box>
  );
};

export default Signup;

```
<img width="414" alt="Screenshot 2024-03-15 at 10 20 21 PM" src="https://github.com/expo/expo/assets/89607202/aba8e518-ae84-4b78-8f7e-285025aabe96">"

Thank you!

**Comment by ganeshkumarburra** [comment]
Thank you @douglowder , For replying quickly.
It took sometime to me do test in the Existing App, there are some issues which i am facing after moving from native-stack to stack.

Then i created one more sample app to test this.
Now, i am able to Load the Fonts in Stack screens too.
But now i have one more challenge?
After moving the Stack Navigation which you mentioned, i am getting below error
ERROR  TypeError: _RNGestureHandlerModule.default.flushOperations is not a function (it is undefined), js engine: hermes
Surprisingly, this issue comes when we click on reload (r), in expo app, but it will go away when we save any other .tsx file.
**App.tsx**
```js
import { StatusBar } from 'expo-status-bar';
import { StyleSheet, Text, View } from 'react-native';
import { createStackNavigator } from '@react-navigation/stack';
import { NavigationContainer } from '@react-navigation/native';
import 'react-native-gesture-handler';
import { useFonts } from 'expo-font';
import * as SplashScreen from 'expo-splash-screen';
import Home from './src/screens/home';
import { useCallback } from 'react';

SplashScreen.preventAutoHideAsync();
const Stack = createStackNavigator();

export default function App() {

  const [fontsLoaded, fontError] = useFonts({
    'LemonLove': require('./assets/fonts/LemonLove.ttf'),
  });

  const onLayoutRootView = useCallback(async () => {
    if (fontsLoaded || fontError) {
      await SplashScreen.hideAsync();
    }
  }, [fontsLoaded, fontError]);

  if (!fontsLoaded && !fontError) {
    return null;
  }
  
  return (
    <NavigationContainer>
      <Stack.Navigator>
        <Stack.Screen name="Home" component={Home} />
      </Stack.Navigator>
    </NavigationContainer>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
});

```
**Home.tsx**
```js
import { StatusBar } from 'expo-status-bar';
import { StyleSheet, Text, View } from 'react-native';
export default function Home() {
    return (
      <View >
        <Text style={{fontFamily:"LemonLove"}}>Home Screen</Text>
      </View>
    );
  }
```
**package.json**
```json
{
  "name": "lecturepedia",
  "version": "1.0.0",
  "main": "node_modules/expo/AppEntry.js",
  "scripts": {
    "start": "expo start",
    "android": "expo start --android",
    "ios": "expo start --ios",
    "web": "expo start --web",
    "ts:check": "tsc"
  },
  "dependencies": {
    "@react-navigation/native": "^6.1.16",
    "@react-navigation/stack": "^6.3.28",
    "expo": "~50.0.13",
    "expo-font": "^11.10.3",
    "expo-splash-screen": "^0.26.4",
    "expo-status-bar": "~1.11.1",
    "react": "18.2.0",
    "react-native": "0.73.5",
    "react-native-gesture-handler": "^2.15.0",
    "react-native-safe-area-context": "^4.9.0"
  },
  "devDependencies": {
    "@babel/core": "^7.20.0",
    "@types/react": "~18.2.45",
   "typescript": "^5.1.3"
  },
  "private": true
}
```

<img width="414" alt="Screenshot 2024-03-16 at 12 09 31 AM" src="https://github.com/expo/expo/assets/89607202/1bd021e3-d4cc-4d5b-8d44-5da8dac29468">
<img width="414" alt="Screenshot 2024-03-16 at 12 09 47 AM" src="https://github.com/expo/expo/assets/89607202/6a03f331-564c-4c20-9512-62803811a919">

**Comment by douglowder** [comment]
I believe you are supposed to import GestureHandlerRootView from the gesture handler package, and then wrap your app in the GestureHandlerRootView component. Example here: https://github.com/infinitered/ignite/blob/master/boilerplate/app/app.tsx

**Comment by chasem-dev** [comment]
I noticed this happening today, on both SDK 50 and SDK 51. Get a false and null for fontsLoaded and fontError, but never get a response update after that, and I don't see anything in my console to explain what's going on. Seems like it's infinitely loading, or silently failing

**Comment by brentvatne** [comment]
@chasem-dev - can you share a minimal reproducible example in a new issue?

**Comment by chasem-dev** [comment]
I apologize @brentvatne updated my comment just now, found a bug in my repo causing it 🤦

**Comment by slauzinho** [comment]
I'm having a similar issue using sdk 51 and expo-font `12.0.4` where the font initially loads fine, but the next time I re-open the app the fonts are no longer loaded.

**Comment by GaeDLLD** [comment]
Same issue as you @pera14 and @slauzinho... Since I update my project to Expo 51 with react-native 0.74 and expo-font 12.04... My custom .otf font is loading and displaying as requested on the first dev build (but with the same dismissible error as @pera14). And when I fast-reload my app, the text loose the custom font family and all others text styles...

**Comment by slauzinho** [comment]
I tried upgrading expo-font to `12.0.5` as now instead of showing a default font tthe app just crashes and I get the log:
```
EXC_BAD_ACCESS: elvetica Neue Interface UltraLight P2 >
Attempted to dereference null **pointer.**
```

**Comment by pera14** [comment]
I did following and it's working for me now.
1. I added `expo-font` and added following in `app.json`:
```json
{
	"expo": {
		...
		"plugins": [
			...
			[
				"expo-font",
				{
					"fonts": [
						"./assets/fonts/nunito-semi-bold.ttf", 
						"./assets/fonts/nunito-medium.ttf", 
						"./assets/fonts/nunito-bold.ttf", 
						"./assets/fonts/nunito-classic.ttf", 
						"./assets/fonts/nunito.ttf"]
				}
			]
		],
    }
}
```

2. I used `useFonts`: 
```js
const [fontsLoaded, fontError] = useFonts({
		'nunito-semi-bold': require('./assets/fonts/nunito-semi-bold.ttf'),
		'nunito-medium': require('./assets/fonts/nunito-medium.ttf'),
		'nunito-bold': require('./assets/fonts/nunito-bold.ttf'),
		'nunito-classic': require('./assets/fonts/nunito-classic.ttf'),
		nunito: require('./assets/fonts/nunito.ttf'),
	});
```
But I start the app even if an error occurs, since the font will be loaded in runtime. This solution works for me so far.

**Comment by GaeDLLD** [comment]
> I did following and it's working for me now.
> 
> 1. I added `expo-font` and added following in `app.json`:
> 
> ```json
> {
> 	"expo": {
> 		...
> 		"plugins": [
> 			...
> 			[
> 				"expo-font",
> 				{
> 					"fonts": [
> 						"./assets/fonts/nunito-semi-bold.ttf", 
> 						"./assets/fonts/nunito-medium.ttf", 
> 						"./assets/fonts/nunito-bold.ttf", 
> 						"./assets/fonts/nunito-classic.ttf", 
> 						"./assets/fonts/nunito.ttf"]
> 				}
> 			]
> 		],
>     }
> }
> ```
> 
>  2. I used `useFonts`:
> 
> ```js
> const [fontsLoaded, fontError] = useFonts({
> 		'nunito-semi-bold': require('./assets/fonts/nunito-semi-bold.ttf'),
> 		'nunito-medium': require('./assets/fonts/nunito-medium.ttf'),
> 		'nunito-bold': require('./assets/fonts/nunito-bold.ttf'),
> 		'nunito-classic': require('./assets/fonts/nunito-classic.ttf'),
> 		nunito: require('./assets/fonts/nunito.ttf'),
> 	});
> ```
> 
> But I start the app even if an error occurs, since the font will be loaded in runtime. This solution works for me so far.

This is what I already did, but still not working after refresh. And custom font not working at all on Android...

**Comment by anstap** [comment]
Have the same issue using Icomoon font and did the steps in the reply above no luck, still fails. It seems to be failing in production build uncontrollably and on the iOS simulator when the hot-refresh happens.

**Comment by vbylen** [comment]
happens on every reload of the simulator

<img width="353" alt="image" src="https://github.com/expo/expo/assets/57633368/8e890c54-e335-4300-ac50-52d06e209a3a">

**Comment by aleqsio** [comment]
@vbylen I think this PR should fix the error 203 issue:
https://github.com/expo/expo/pull/28989

**Comment by davidfarinha** [comment]
Still happening for me on the latest version. Did anyone find a solution?

**Comment by AliBahaari** [comment]
I'm using version 51.0.28 and still have the same problem.

**Comment by outaTiME** [comment]
I’ve been dealing with random font issues on Android for a while now, and I still haven’t found a solution or seen any reports about it. 

On certain devices and in specific situations, the app sometimes crashes on a cold start with an error from `expo.modules.font.FontLoaderModule$definition$lambda$2$$inlined$AsyncFunction$3.invoke`. 

I keep seeing tons of these messages in Google Play Console.

**Comment by outaTiME** [comment]
> I’ve been dealing with random font issues on Android for a while now, and I still haven’t found a solution or seen any reports about it.
> 
> On certain devices and in specific situations, the app sometimes crashes on a cold start with an error from `expo.modules.font.FontLoaderModule$definition$lambda$2$$inlined$AsyncFunction$3.invoke`.
> 
> I keep seeing tons of these messages in Google Play Console.

I’m using `expo-font` in the following way:

```
plugins: [
  [
    'expo-font',
    {
      fonts: ['./assets/fonts/FiraGO-Regular.otf'],
    },
  ],
]
```

```
const App = () => {
  const [fontsLoaded, fontsError] = useFonts({
    ...MaterialIcons.font,
    ...MaterialCommunityIcons.font,
    ...FontAwesome6.font,
  });
  const appIsLoading = !(fontsLoaded || fontsError);
  // trace when fonts load error
  if (fontsError) {
    console.warn('Error while trying to load fonts', fontsError);
  }
  if (appIsLoading) {
    return null;
  }
  return <View />;
};
```

**Comment by enzo-perez-itti** [comment]
Im having the same issue on ios in debug

1) Kill the app
2) open again
3) the fonts doesn't work
4) do some change in code and save
5) the fonts start working

 between the steps 1 and 2, if I add 
 ```
 if (!loaded && !error) {
    return <ActivityIndicator />; //or return null or whatever
  }
  ```
 
  The fonts works ok but the app shows my home screen for one sec and the back to the login (correct screen)

**Comment by LeaFuryk** [comment]
I have same issue as @enzo-perez-itti 

```
  const { isLoading } = useSession()
  const { fontsLoaded } = useFontsLoader()

  useEffect(() => {
    console.log("fontsLoaded: ",fontsLoaded)
    console.log("isLoading: ", isLoading)
    if (fontsLoaded && !isLoading) {
      await SplashScreen.hideAsync()
    }
  }, [fontsLoaded, isLoading])

  if (!fontsLoaded && isLoading) {
    return null
  }
```
Logs:
```
 LOG  fontsLoaded:  false
 LOG  isLoading:  false
```
Error from `useFonts` hook is null btw

Same sprint I installed `expo-updates` on the project this started to happen 🤔

**Comment by jeffngugi** [comment]
Any new solution for this? 
expo 51 keeps randomly failing in iOS devices

**Comment by Wizzel1** [comment]
I have the same issue, the `fontsLoaded` is true but in one particular component, the font does not get applied.

**Comment by Tranthanh98** [comment]
> I'm having a similar issue using sdk 51 and expo-font `12.0.4` where the font initially loads fine, but the next time I re-open the app the fonts are no longer loaded.

same issue

**Comment by Tranthanh98** [comment]
> Any new solution for this? expo 51 keeps randomly failing in iOS devices

same, It works for the first several time, but then no works
